packages/material-ui/src/MenuItem/MenuItem.js

```diff
@@ -2,87 +2,201 @@ import * as React from 'react';
+      theme.palette.action.selectedOpacity + theme.palette.action.hoverOpacity,
+    ),
+    // Reset on touch devices, it doesn't add specificity
+    '@media (hover: none)': {
+      backgroundColor: alpha(theme.palette.primary.main, theme.palette.action.selectedOpacity),
+    },
+  },
+  [`&.${menuItemClasses.focusVisible}`]: {
+    backgroundColor: theme.palette.action.focus,
+  },
+  [`&.${menuItemClasses.disabled}`]: {
+    opacity: theme.palette.action.disabledOpacity,
+  },
+  [`& + .${dividerClasses.root}`]: {
+    marginTop: theme.spacing(1),
+    marginBottom: theme.spacing(1),
+  },
+  [`& + .${dividerClasses.inset}`]: {
+    marginLeft: 52,
+  },
+  [`& .${listItemTextClasses.root}`]: {
+    marginTop: 0,
+    marginBottom: 0,
+  },
+  [`& .${listItemTextClasses.inset}`]: {
+    paddingLeft: 36,
   },
+  [`& .${listItemIconClasses.root}`]: {
+    minWidth: 36,
+  },
+  ...(!styleProps.dense && {
+    [theme.breakpoints.up('sm')]: {
+      minHeight: 'auto',
+    },
+  }),
   ...(styleProps.dense && {
+    minHeight: 36,
     ...theme.typography.body2,
-    minHeight: 'auto',
+    [`& .${listItemIconClasses.root} svg`]: {
+      fontSize: '1.25rem',
+    },
   }),
 }));
 
 const MenuItem = React.forwardRef(function MenuItem(inProps, ref) {
   const props = useThemeProps({ props: inProps, name: 'MuiMenuItem' });
   const {
-    className,
+    autoFocus = false,
     component = 'li',
     dense = false,
+    divider = false,
     disableGutters = false,
-    ListItemClasses,
+    focusVisibleClassName,
     role = 'menuitem',
-    selected,
     tabIndex: tabIndexProp,
     ...other
   } = props;
 
-  const styleProps = { dense };
+  const context = React.useContext(ListContext);
+  const childContext = {
+    dense: dense || context.dense || false,
+    disableGutters,
+  };
+
+  const menuItemRef = React.useRef(null);
+  useEnhancedEffect(() => {
+    if (autoFocus) {
+      if (menuItemRef.current) {
+        menuItemRef.current.focus();
+      } else if (process.env.NODE_ENV !== 'production') {
+        console.error(
+          'Material-UI: Unable to set focus to a MenuItem whose component has not been rendered.',
+        );
+      }
+    }
+  }, [autoFocus]);
+
+  const styleProps = {
+    ...props,
+    dense: childContext.dense,
+    divider,
+    disableGutters,
+  };
 
   const classes = useUtilityClasses(props);
 
+  const handleRef = useForkRef(menuItemRef, ref);
+
   let tabIndex;
   if (!props.disabled) {
     tabIndex = tabIndexProp !== undefined ? tabIndexProp : -1;
   }
 
   return (
-    <ListItem
-      components={{ Root: MenuItemRoot }}
-      componentsProps={{ root: { styleProps } }}
-      button
-      role={role}
-      tabIndex={tabIndex}
-      component={component}
-      selected={selected}
-      disableGutters={disableGutters}
-      className={clsx(classes.root, className)}
-      ref={ref}
-      {...other}
-      classes={ListItemClasses}
-    />
+    <ListContext.Provider value={childContext}>
```

Comment:
Do we still need to use the `ListContext`?

Reply:
Yes, there is one case I that see notice.

```tsx
<MenuItem dense>
  <ListItemText>Primary text</ListItemText>
</MenuItem>
```

I think let's keep the same behavior as v4.